### PR TITLE
Currency exchange

### PR DESCRIPTION
**Pull request by xeric**
## Description

Support latest currency code and country mapping file, fix a issue in extension which will cause undefined object access exception.

## Screencast

No Screenshot since a small logic change

## Checklist

- [√] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [√] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [√] I ran `npm run build` and [tested this distribution build in Raycast](https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration)
- [√] I checked that files in the `assets` folder are used by the extension itself
- [√] I checked that assets used by the `README` are placed outside of the `metadata` folder

### CODE REVIEW

File: extensions/currency-exchange/src/index.tsx [hunk 1]
```diff
@@ -0,0 +1,425 @@
...
+  return currencyData as CurrencyResult;
+}
+
+function getFlagEmoji(countryCode: string): string {
+  const codePoints = countryCode
+    .toUpperCase()
+    .split("")
+    .map((char) => 127397 + char.charCodeAt(0));
+  let v = countryCode.startsWith("X") ? "🇺🇳" : String.fromCodePoint(...codePoints) || "🇺🇳";
+  v = countryCode === "TW" ? "🇨🇳" : v;
```
Comment (pan93412) [inline extensions/currency-exchange/src/index.tsx:395]: I don't think that this replacement is necessary. 🤔

Why not just leave it as what the Unicode defines; or add a configuration entry, to allow users setting what flags to replace? (It should not be hardcoded.)

**Review by pan93412 — CHANGES_REQUESTED** [review]

File: extensions/currency-exchange/src/index.tsx (hunk 1 above)
Comment (wdzeng) [reply extensions/currency-exchange/src/index.tsx:395]: 支那人XDD

Comment (kuohuanhuan) [reply extensions/currency-exchange/src/index.tsx:395]: 偷塞私貨 XD

**Comment by brlin-tw** [comment]
Dear organization/project maintainers(@dehesa, @fe9lix, @itsnwa, @mathieudutour, @PitNikola, and @thomaspaulmann), apologies for the nuisance but I would like to escalate an issue that is alarmingly urgent in this PR.

The author of this PR purposely include a change in commit ff9d045dfe8ff9590ed6a13535f3702b19967e0a that contains a _highly politically incorrect_ country flag replacement for Taiwan at [extensions/index.tsx at ff9d045, line 371](https://github.com/raycast/extensions/blob/ff9d045dfe8ff9590ed6a13535f3702b19967e0a/extensions/currency-exchange/src/index.tsx#L371).  This indicates that the author _may_ impose other potentially harmful change that fits towards dictatorship China government's agenda.  I strongly urge you to request the PR author to drop the problematic change and to give this PR a **thorough security review** to avoid possible consequences.

Suggestions:

* Use/Keep the in facto Taiwan country flag(which is the Republic of China flag), if China users require a different flag due to local regulation requirements, special-case it and make it an opt-in "Make China happy" option.
* Avoid the country flag interface at all as it is a political minefield for countries that have sovereignty controversies.

Thank you.

## Reference

* [Political status of Taiwan - Wikipedia](https://en.wikipedia.org/wiki/Political_status_of_Taiwan)

**Comment by brlin-tw** [comment]
Note that the PR author also [silently resolves the change request thread](https://github.com/raycast/extensions/pull/2861#pullrequestreview-1103239155) from the Taiwanese/pro-Taiwan reviewer(s), which is a bad sign indicating that they may want to hide the controversy under the carpet.

**Comment by brlin-tw** [comment]
The commit 92e5cca introduces an opt-out toggle which is the _exact opposite_ of what I suggested at https://github.com/raycast/extensions/pull/2861#issuecomment-1243333017, and **will harm all Raycast users that are not intend to support the One-China policy BS to face public relation danger**.

**Comment by brlin-tw** [comment]
The commit 0b952ad drops the entire TWD currency that is subjectively widely in use in the field.  Please stop wasting everyone's time and make an opt-in toggle to make your government happy without dragging the rest of the world into your government's agenda.  Thank you.

**Comment by xeric** [comment]
> The commit [0b952ad](https://github.com/raycast/extensions/commit/0b952ad4a10e45feadbfff0c7753dee48b8e94e0) drops the entire TWD currency that is subjectively widely in use in the field. Please stop wasting everyone's time and make an opt-in toggle to make your government happy without dragging the rest of the world into your government's agenda. Thank you.
> 
> "Conforming to China government's country flag requirements" would be a sensible toggle name, my 2 cents.

You liked to make someone happy, you thought it is an option, I provided an option to make Taiwan happy, you guys are unhappy.
I think I'd better to totally remove some feature to avoid arguments, you guys are unhappy.
Why you guys just say "hey, administrator, force this guy to do what we like" directly?
If you do not allow different voices, and do not allow silence, think about what role you are, and how you guys dare to say "I'm neutral", do you feel funny?
Again, I did not say this extension will support every country and region, especially this feature will lead arguments, no matter true or false will have someone has concern.

**Comment by brlin-tw** [comment]
> I provided an option to make Taiwan happy, you guys are unhappy.

I've already made a reasonable argument to make the China government preferred flags *opt-in*, instead of opt-out.  You made the exact opposite which of course will make us unhappy.

> I think I'd better to totally remove some feature to avoid arguments, you guys are unhappy.

If you simply drop the controversial feature alone we won't have any word about it.

> Why you guys just say "hey, administrator, force this guy to do what we like" directly?

Because it's totally not nice to sneak a controversial change that presents your own government's idealogy without asking the project maintainer and other stakeholders whether they like it or not.  Escalating the issue for closer inspection is a reasonable reaction to such behavior.

> If you do not allow different voices

Your comment button is not stuck, my darling.

> how you guys dare to say "I'm neutral"

The only being that says the N-word in this thread is you, we never claimed to be such a thing.

> I did not say this extension will support every country and region

You never claim such a thing until the latest revision, we can reasonably believe that it should support every country and region.

---

Not going to let this thread lock with `too heated` tag so this will be my last argument, I'll respectfully leave the matter to the project maintainers to progress the PR in the right direction.  Cheers :clinking_glasses: .

**Comment by pan93412** [comment]
@xeric Your app is great – I appreciate to your contribution to Raycast ecosystem, and I want to propose a suggestion.

I think there is another way: **enable the replacement by default if the user is in China and other regions; and leave it opt-in for others.** Similar to what Apple does, it matches the policy in China when using there, while still respects people in Taiwan and elsewhere.

I am not sure if Raycast provides any API to accomplish it, and I wish you can find a way that makes everyone happy ;)

**Comment by pan93412** [comment]
Besides:

1. It is better to name the option more explicitly. Nobody can understand “make Taiwan happy” easily, and this name is a bit controversial. I suggest to name it as “Using another standard of flags,” or “Follow China's policy of region flags,” which is more explicit while makes everyone happy.

2. The commit log of this branch is not really clear. It is better to clean up your branch, and squashing them to one commit is a good way to do. I did not test the following commands, and you are welcome to ask me if something wrong when running these commands :)

        # Reset the main branch.
        git checkout main
        git fetch origin main
        git reset --hard FETCH_HEAD
        
        # Checkout another branch.
        git checkout -b currency-exchange-2
        
        # The tree is clean; only the commit log
        # is out of order. Let's squash them to a
        # commit.
        git merge --squash currency-exchange
        
        # Force-push to the remote branch.
        git push -f [your_remote] HEAD:currency-exchange

**Comment by xeric** [comment]
> @xeric Your app is great – I appreciate to your contribution to Raycast ecosystem, and I want to propose a suggestion.
> 
> I think there is another way: **enable the replacement by default if the user is in China, and leave it opt-in for others.** Similar to what Apple does, it matches the policy in China when using there, while still respects people in Taiwan and elsewhere.
> 
> I am not sure if Raycast provides any API to accomplish it, and I wish you can find a way that makes everyone happy ;)

Unfortunately, you are the only one who is calm enough to talk about features, not political, I will to keep communication with you regard this feature, not like someone comments "N-word is same case" or "Chi Na", someone want to get respects from others, first of all, provide respects (TBH, I did not plan to hide something in an open source project).

I planned to response reply those unreasonable comments but I think it may help nothing except increase arguments, so will not anymore.

I agree to have a transparent option if Raycast can provide some API, and maybe before that I will let the flag configurable in "command option".

Yes, you are right, the commits contained some additional useless history from my another extension, will do a clean up.

**Comment by wdzeng** [comment]
> Unfortunately, you are the only one who is calm enough to talk about features, not political.

Actually, you are the very first person to be political, who tries to serve your own Great government.

> someone want to get respects from others, first of all, provide respects

You change other's flag. Is this behavior considered respectful?

> TBH, I did not plan to hide anything in an open-source project

You did. You tried to hide the One-China policy in this PR. If @pan93412 or @brlin-tw did not point out the code I guess no one would find it.

**Comment by thomaspaulmann** [comment]
Hey,

Sorry for being a bit late to the discussion. Generally, we trust the author to act in good faith to update their extensions. Here, it seems that the changes have affected users of the extension. So, let’s try to resolve this contribution.

The ExchangeRate-API documents a list of [supported currencies](https://exchangerate-api.com/docs/supported-currencies), including the New Taiwan Dollar (TWD). The extension should display all supported currencies to guarantee full functionality.

The extension can be adjusted if local government policies affect the list of supported currencies. We aim to keep manual configurations of extensions to a minimum. Ideally, we can find a solution that “just works”, as suggested by @pan93412. We don’t provide such an API, but a user’s locale (`Intl.DateTimeFormat().resolvedOptions().locale`) or a geolocation API might be sufficient.

A few concrete suggestions:
- Bring back support for TWD
- Automatically detect country to apply potential local policies
- Update the README accordingly

I hope we can alter the extension to work for all users out-of-the-box.

**Comment by xeric** [comment]
> 

Thanks @thomaspaulmann, locale would not be accurate value for making decision, geo API would impact performance somehow.
I finally remove the mapping code and rename all variables from country to countryAndRegion for avoiding any arguments.

File: extensions/currency-exchange/src/index.tsx [hunk 2]
```diff
@@ -226,7 +226,6 @@ function useExchange() {
         isLoading: true,
       }));
       try {
-        console.log("country:", state.fromCurrencyCode);
```
Comment (Brli) [inline extensions/currency-exchange/src/index.tsx:229]: What's wrong with this conole.log line?

Comment (pan93412) [reply extensions/currency-exchange/src/index.tsx:229]: I think it is just the a log for debugging, and it indeed should be deleted when publishing to production.

File: extensions/currency-exchange/src/index.tsx [hunk 3]
```diff
@@ -384,14 +387,13 @@ function enrichExchangeData(
...
-  const codePoints = countryCode
+function getFlagEmoji(countryAndRegionCode: string): string {
+  const codePoints = countryAndRegionCode
     .toUpperCase()
     .split("")
     .map((char) => 127397 + char.charCodeAt(0));
-  let v = countryCode.startsWith("X") ? "🇺🇳" : String.fromCodePoint(...codePoints) || "🇺🇳";
-  v = countryCode === "TW" ? "🇨🇳" : v;
-  v = countryCode === "AN" ? "🇺🇳" : v;
+  let v = countryAndRegionCode.startsWith("X") ? "🇺🇳" : String.fromCodePoint(...codePoints) || "🇺🇳";
```
Comment (Artoria2e5) [inline extensions/currency-exchange/src/index.tsx:395]: Not to start another flame war, but XK (Kosovo) usually has its own emojis. It won’t be a problem for currency, but as a general code-to-emoji function it ain’t right.

File: extensions/currency-exchange/src/index.tsx [hunk 4]
```diff
@@ -384,14 +387,13 @@ function enrichExchangeData(
...
+  v = countryAndRegionCode === "AN" ? "🇺🇳" : v;
```
Comment (Artoria2e5) [inline extensions/currency-exchange/src/index.tsx:396]: I am pretty sure penguins will not come here and talk about their flag, but what’s wrong with using the 🇦🇶 flag? In fact, all of this extra processing is pointless.

**Comment by pernielsentikaer** [comment]
I'm a bit lost in this thread, is everything as it should be now, or are there still remaining issues that still need attention?

File: extensions/currency-exchange/src/index.tsx (hunk 3 above)
Comment (xeric) [reply extensions/currency-exchange/src/index.tsx:395]: XK is not in currency support list.

**Comment by xeric** [comment]
> I'm a bit lost in this thread, is everything as it should be now, or are there still remaining issues that still need attention?

Yes, except anyone else raises other concerns.

**Review by goodjack — APPROVED** [review]

**Comment by pernielsentikaer** [comment]
What about you @brlin-tw / @pan93412

**Comment by pan93412** [comment]
> What about you @brlin-tw / @pan93412 

I've approved. LGTM.

**Review by pernielsentikaer — APPROVED** [review]
Hi 👋 

Looks good to me, approved 🔥

**Pull request merged**